File: framing.go
package noise

import (
	"encoding/binary"
	"fmt"
	"io"
)

// writeFrame writes data to w with a 2-byte big-endian length prefix.
func writeFrame(w io.Writer, data []byte) error {
	if len(data) > maxMessageSize {
		return fmt.Errorf("noise: frame too large: %d bytes", len(data))
	}
	header := make([]byte, 2)
	binary.BigEndian.PutUint16(header, uint16(len(data)))
	if _, err := w.Write(header); err != nil {
		return fmt.Errorf("noise: write frame header: %w", err)
	}
	if _, err := w.Write(data); err != nil {
		return fmt.Errorf("noise: write frame body: %w", err)
	}
	return nil
}

// readFrame reads a 2-byte big-endian length prefix then the body from r.
func readFrame(r io.Reader) ([]byte, error) {
	header := make([]byte, 2)
	if _, err := io.ReadFull(r, header); err != nil {
		return nil, fmt.Errorf("noise: read frame header: %w", err)
	}
	size := binary.BigEndian.Uint16(header)
	body := make([]byte, size)
	if _, err := io.ReadFull(r, body); err != nil {
		return nil, fmt.Errorf("noise: read frame body: %w", err)
	}
	return body, nil
}

// nopCloser wraps an io.ReadWriter that doesn't implement io.Closer.
type nopCloser struct{ io.ReadWriter }

func (nopCloser) Close() error { return nil }

File: rust/src/framing.rs
use std::io::{self, Read, Write};

use crate::cipher::MAX_MESSAGE_SIZE;

/// Writes data to w with a 2-byte big-endian length prefix.
pub(crate) fn write_frame(w: &mut dyn Write, data: &[u8]) -> io::Result<()> {
    if data.len() > MAX_MESSAGE_SIZE {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("noise: frame too large: {} bytes", data.len()),
        ));
    }
    let len = data.len() as u16;
    w.write_all(&len.to_be_bytes())?;
    w.write_all(data)?;
    Ok(())
}

/// Reads a 2-byte big-endian length prefix then the body from r.
pub(crate) fn read_frame(r: &mut dyn Read) -> io::Result<Vec<u8>> {
    let mut header = [0u8; 2];
    r.read_exact(&mut header)?;
    let size = u16::from_be_bytes(header) as usize;
    let mut body = vec![0u8; size];
    r.read_exact(&mut body)?;
    Ok(body)
}
